src/Core/MySQLClient.cpp

```diff
@@ -0,0 +1,155 @@
+{
+    in = nullptr;
+    out = nullptr;
+    if (socket)
+        socket->close();
+    socket = nullptr;
+    connected = false;
+}
+
+/// https://dev.mysql.com/doc/internals/en/connection-phase-packets.html
+void MySQLClient::handshake()
+{
+    Handshake handshake;
+    packet_sender->receivePacket(handshake);
+    if (handshake.auth_plugin_name != mysql_native_password)
+    {
+        throw MySQLClientError(
+            "Only support " + mysql_native_password + " auth plugin name, but got " + handshake.auth_plugin_name,
+            ErrorCodes::UNKNOWN_PACKET_FROM_SERVER);
+    }
+
+    Native41 native41(password, handshake.auth_plugin_data);
+    String auth_plugin_data = native41.getAuthPluginData();
+
+    HandshakeResponse handshake_response(
+        client_capability_flags, max_packet_size, charset_utf8, user, "", auth_plugin_data, mysql_native_password);
+    packet_sender->sendPacket<HandshakeResponse>(handshake_response, true);
+
+    PacketResponse packet_response(client_capability_flags, true);
+    packet_sender->receivePacket(packet_response);
+    packet_sender->resetSequenceId();
+
+    if (packet_response.getType() == PACKET_ERR)
+        throw MySQLClientError(packet_response.err.error_message, ErrorCodes::UNKNOWN_PACKET_FROM_SERVER);
+    else if (packet_response.getType() == PACKET_AUTH_SWITCH)
+        throw MySQLClientError("Access denied for user " + user, ErrorCodes::UNKNOWN_PACKET_FROM_SERVER);
+}
+
+void MySQLClient::writeCommand(char command, String query)
+{
+    WriteCommand write_command(command, query);
+    packet_sender->sendPacket<WriteCommand>(write_command, true);
+
+    PacketResponse packet_response(client_capability_flags);
+    packet_sender->receivePacket(packet_response);
+    switch (packet_response.getType())
+    {
+        case PACKET_ERR:
+            throw MySQLClientError(packet_response.err.error_message, ErrorCodes::UNKNOWN_PACKET_FROM_SERVER);
+        case PACKET_OK:
+            break;
+        default:
+            break;
+    }
+    packet_sender->resetSequenceId();
+}
+
+void MySQLClient::registerSlaveOnMaster(UInt32 slave_id)
+{
+    RegisterSlave register_slave(slave_id);
+    packet_sender->sendPacket<RegisterSlave>(register_slave, true);
+
+    PacketResponse packet_response(client_capability_flags);
+    packet_sender->receivePacket(packet_response);
+    packet_sender->resetSequenceId();
+    if (packet_response.getType() == PACKET_ERR)
+        throw MySQLClientError(packet_response.err.error_message, ErrorCodes::UNKNOWN_PACKET_FROM_SERVER);
+}
+
+void MySQLClient::ping()
+{
+    writeCommand(Command::COM_PING, "");
+}
+
+void MySQLClient::startBinlogDump(UInt32 slave_id, String replicate_db, String binlog_file_name, UInt64 binlog_pos)
+{
+    /// Set binlog checksum to CRC32.
+    String checksum = "CRC32";
+    writeCommand(Command::COM_QUERY, "SET @master_binlog_checksum = '" + checksum + "'");
+
+    /// Set heartbeat 30s.
+    UInt64 period_ns = (30 * 1e9);
```

Comment:
MySQL has a number of timeout variables, 30 seconds is likely to exceed their limits and MySQL server will abort the connection, so it is better to set it to 1 seconds to keep alive.